flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/KafkaSource.java

```diff
@@ -178,38 +182,45 @@ private void initializeConnection() {
 		consumerIterator = stream.iterator();
 	}
 
-	/**
-	 * Called to forward the data from the source to the {@link DataStream}.
-	 * 
-	 * @param collector
-	 *            The Collector for sending data to the dataStream
-	 */
 	@Override
-	public void run(Collector<OUT> collector) throws Exception {
-		isRunning = true;
-		try {
-			while (isRunning && consumerIterator.hasNext()) {
-				OUT out = schema.deserialize(consumerIterator.next().message());
-				if (schema.isEndOfStream(out)) {
-					break;
-				}
-				collector.collect(out);
+	public void open(Configuration config) throws Exception {
+		initializeConnection();
+	}
+
+	@Override
+	public boolean reachedEnd() throws Exception {
+		if (nextElement != null) {
+			return true;
+		} else if (consumerIterator.hasNext()) {
+			OUT out = schema.deserialize(consumerIterator.next().message());
+			if (schema.isEndOfStream(out)) {
+				return false;
 			}
-		} finally {
-			consumer.shutdown();
+			nextElement = out;
 		}
+		return false;
 	}
 
 	@Override
-	public void open(Configuration config) throws Exception {
-		initializeConnection();
+	public OUT next() throws Exception {
+		if (nextElement != null) {
+			OUT out = nextElement;
+			nextElement = null;
+			return out;
+		}
+
+		MessageWithMetadata msg;
+		OUT out = schema.deserialize(consumerIterator.next().message());
+
+		if (schema.isEndOfStream(out)) {
```

Comment:
Why are you throwing an exception here?
The old code stopped reading on that event, without failing